test/jdk/javax/swing/JEditorPane/4325606/bug4325606.java

```diff
@@ -0,0 +1,132 @@
+ * questions.
+ */
+
+/* @test
+ * @bug 4325606
+ * @summary Tests getting row start
+ * @key headful
+ * @run main bug4325606
+ */
+
+import java.awt.AWTException;
+import java.awt.BorderLayout;
+import java.awt.Robot;
+import java.awt.event.InputEvent;
+import java.awt.event.MouseAdapter;
+import java.awt.event.MouseEvent;
+import java.awt.event.WindowAdapter;
+import java.awt.event.WindowEvent;
+import java.lang.reflect.InvocationTargetException;
+import javax.swing.JEditorPane;
+import javax.swing.JFrame;
+import javax.swing.JScrollPane;
+import javax.swing.SwingUtilities;
+import javax.swing.text.BadLocationException;
+import javax.swing.text.Utilities;
+
+public class bug4325606 {
+
+    public volatile boolean passed = true;
+    private JFrame frame;
+    private JEditorPane pane;
+
+    public void setupGUI() {
+        frame = new JFrame("Click Bug");
+        frame.setLayout(new BorderLayout());
+
+        pane = new JEditorPane();
+        pane.addMouseListener(new ClickListener());
+        pane.setContentType("text/html");
+        pane.setText("<html><body>" +
+                "<p>Here is line one</p>" +
+                "<p>Here is line two</p>" +
+                "</body></html>");
+
+        frame.add(new JScrollPane(pane), BorderLayout.CENTER);
+
+        frame.addWindowListener(new TestStateListener());
+        frame.setLocation(50, 50);
+        frame.setSize(400, 300);
+        frame.setVisible(true);
+    }
+
+    class TestStateListener extends WindowAdapter {
+        public void windowOpened(WindowEvent ev) {
+            Robot robo;
+            try {
+                robo = new Robot();
+            } catch (AWTException e) {
+                throw new RuntimeException("Robot could not be created");
+            }
+            robo.setAutoDelay(100);
+            robo.delay(1000);
+            robo.mouseMove(100,100);
+            robo.mousePress(InputEvent.BUTTON1_DOWN_MASK);
+            robo.mouseRelease(InputEvent.BUTTON1_DOWN_MASK);
+            robo.mousePress(InputEvent.BUTTON1_DOWN_MASK);
+            robo.mouseRelease(InputEvent.BUTTON1_DOWN_MASK);
+            robo.mousePress(InputEvent.BUTTON1_DOWN_MASK);
+            robo.mouseRelease(InputEvent.BUTTON1_DOWN_MASK);
+        }
+    }
+
+    class ClickListener extends MouseAdapter {
+        public void mouseClicked(MouseEvent event) {
+            try {
+                Utilities.getRowStart(pane, pane.getCaretPosition());
+            } catch (BadLocationException blex) {
+                passed = false;
+            }
+        }
+    }
+
+    public void cleanupGUI() {
+        if (frame != null) {
+            frame.setVisible(false);
+            frame.dispose();
+        }
+    }
+
+    public static void main(String[] args) throws InterruptedException,
+            InvocationTargetException {
+        bug4325606 b = new bug4325606();
+        try {
+            SwingUtilities.invokeAndWait(b::setupGUI);
+            safeSleep(3000);
```

Comment:
Instead of sleeping, you can use a `CountDownLatch(3)` which you'll `countDown()` for each mouse click. Here you'll call `await(3, TimeUnit.SECONDS)` and throw a timeout error if `await` returns `false`.

Another synchroniser may be used to handle the case where `BadLocationException` is thrown to fail the test right away. Alternatively, `BadLocationException` may be wrapped into `RuntimeException` and re-thrown.

Reply:
This is a very simple case and sleeping worked for it so i do not see reason to rewrite it with CDL.

Reply:
Using CDL can reduce the execution time… likely the three clicks are handled more quickly than in 3 seconds.

Reply:
On the other hand, overengineering a simple case could do more harm… as using `CountDownLatch` isn't straightforward in this case.